### PR TITLE
correctly lower `impl const` to bind to host effect param

### PR DESCRIPTION
**Pull request by fee1-dead**
r? @oli-obk 

cc https://github.com/rust-lang/rust/issues/110395

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs [hunk 1]
```diff
@@ -1349,7 +1349,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         }
                     }
                     GenericParamDefKind::Const { has_default } => {
-                        if !infer_args && has_default {
+                        if !infer_args
+                            && has_default
+                            && !tcx.has_attr(param.def_id, sym::rustc_host)
```
Comment (oli-obk) [inline compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:1354]: How does this work for desugared params?

Comment (fee1-dead) [reply compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:1354]: desugared params get this attribute as well: [link](https://github.com/rust-lang/rust/pull/114545/files#diff-0a61b538a3cec072c76fecae4635af6a12ec3256860029ac70549c2aa53ab394L1433-R1470)

**Comment by oli-obk** [comment]
r=me with tidy happy

**Comment by fee1-dead** [comment]
> r=me with tidy happy

a review of [this commit](https://github.com/rust-lang/rust/pull/114545/commits/8d91309513d83e16771bc87560ef85e10aa3632f) which fixes the ICE would be nice

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 2]
```diff
@@ -1358,30 +1358,55 @@ fn impl_trait_ref(
...
 }
 
-fn check_impl_constness(tcx: TyCtxt<'_>, is_const: bool, ast_trait_ref: &hir::TraitRef<'_>) {
+fn check_impl_constness(
+    tcx: TyCtxt<'_>,
+    is_const: bool,
+    ast_trait_ref: &hir::TraitRef<'_>,
+) -> Option<ErrorGuaranteed> {
     if is_const {
         if let Some(trait_def_id) = ast_trait_ref.trait_def_id() && !tcx.has_attr(trait_def_id, sym::const_trait) {
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/collect.rs:1397]: nit: you can now `?` away the value instead of using `if let Some`

Comment (fee1-dead) [reply compiler/rustc_hir_analysis/src/collect.rs]: Done and removed rightward drift.

File: compiler/rustc_hir_analysis/src/collect.rs [hunk 3]
```diff
@@ -1358,30 +1358,55 @@ fn impl_trait_ref(
...
+                &ast_trait_ref,
+            ) {
+                // we have a const impl, but for a trait without `#[const_trait]`, so
+                // without the host param. If we continue with the HIR trait ref, we get
+                // ICEs for generic arg count mismatch. We do a little HIR editing to
+                // make astconv happy.
+                let mut path_segments = ast_trait_ref.path.segments.to_vec();
+                let last_segment = path_segments.len() - 1;
+                let mut args = path_segments[last_segment].args().clone();
+                args.args = &args.args[..args.args.len() - 1];
```
[the remark this answers is not in the document]
Comment (fee1-dead) [reply compiler/rustc_hir_analysis/src/collect.rs:1377]: This turns out to be a bit more complicated and I didn't want to look inside bodies, so I just added the attr to all automatically generated bodies and asserted that the attr exists instead.

**Comment by fee1-dead** [comment]
spurious

@bors retry

**Pull request merged**